File: Examples/Deprecated/Archive-Job-Execution-Outputs/Perl/exportJobHistory.pl
use LWP::UserAgent;
use HTTP::Cookies;
use JSON;
use JSON::RPC::Client2;
use URI::Escape;
use Data::Dumper;
use POSIX qw(strftime);
use Try::Tiny;
use Config::Tiny;
use DateTime;
use File::Basename 'basename', 'dirname';

### Initial authentication
my $cookie_jar = HTTP::Cookies->new( ignore_discard => 1 );
my $ua = LWP::UserAgent->new(
    cookie_jar => $cookie_jar,
    ssl_opts   => { SSL_use_cert => 0, verify_hostname => 0 },
#    debug => 1 
);

my $HOST = 'host';
my $USERNAME = 'username';
my $PASSWORD = 'password';
my $LAST_SESSION_END = 'lastSessionEnd';
	
sub usage_and_exit
{
	die "Usage:\n  exportTermLogs.pl [ <output file> ] [ <config file> ]\n";
}

sub main
{
	my ($output_file, $config_file);
	
	if ($#ARGV == -1)
	{
		$output_file = "JobHistory" . (strftime "%y%m%d-%H%M%S", localtime) . ".txt";
		$config_file = './config.ini';
	}
	elsif ($ARGV[1] eq '')
	{
		usage_and_exit();
	}
	else
	{
		$output_file = $ARGV[0];
		$config_file = $ARGV[1];
	}
	
	my ($host, $user, $password, $ini_last_end_time);

	my $config = Config::Tiny->read($config_file);

	foreach my $section (keys %{$config}) 
	{
		$host = $config->{$section}->{$HOST};
	    $user = $config->{$section}->{$USERNAME};
	    $password = $config->{$section}->{$PASSWORD};
	
	    if($config->{$section}->{$LAST_SESSION_END})
	    {
	    	$ini_last_end_time = $config->{$section}->{$LAST_SESSION_END};
	    }
	    else
	    {
	    	$ini_last_end_time = 0;
	    }
	    
	    try
	    {
	    	$ini_last_end_time = export_job_history($output_file, $host, $user, $password, $ini_last_end_time);
	    	$config->{$section}->{$LAST_SESSION_END} = $ini_last_end_time if ($ini_last_end_time);
	    }
	    catch
	    {
	    	warn "caught error: $_";
	    }
	}

	$config->write($config_file);
}

sub millis_to_date
{
	my $millis = shift;
    my $tmp = strftime('%Y-%m-%d %H:%M:%S', gmtime($millis/1000));

    return $tmp;
}

sub make_job_list
{
	my $execution = shift;
    my $start_time = $execution->{'startTime'} / 1000;
    my $end_time = $execution->{'endTime'} / 1000;
    
    my $job_list = "\n[ ".$execution->{'status'}." / ".
            $execution->{'jobName'}." / ".
            $execution->{'managedNetwork'}." / ". 
            $execution->{'executor'}." / S:".
            strftime('%Y-%m-%d %H:%M:%S', gmtime($start_time))." / E:".
            strftime('%Y-%m-%d %H:%M:%S', gmtime($end_time)).
            " ]\n";
            
    return $job_list;
}

sub export_job_history
{
	my ($output_file, $host, $user, $password, $ini_last_end_time) = @_;
    my ($start_time, $end_time, $filename, $directory, $params);

    my $netld = JSON::RPC::Client2->new("https://".$host."/rest?j_username=".uri_escape_utf8($user)."&j_password=".uri_escape_utf8($password), $ua);

    my $pageData = $netld->call('Scheduler.getExecutionData', {
	       	pageData => {
	            offset   => 0,
	            pageSize => 1000
	        },
	        sortColumn => 'endTime',
	        descending => 1,
	    }
    );

    foreach my $execution (@{$pageData->{'executionData'}})
    {
        $start_time = millis_to_date($execution->{'startTime'});
        $last_end_time = $execution->{'endTime'};
        $end_time = millis_to_date($last_end_time);

        my $execution_id = $execution->{'id'};

        my $details = $netld->call('Plugins.getExecutionDetails', { 'executionId' => $execution_id });

        if ($details)
        {
			open my $fh, '>>', $output_file or die "$!";
            for my $detail (@$details)
            {
                if ($execution->{'jobType'} eq 'Script Tool Job')
                {
                    if ($last_end_time > $ini_last_end_time)
                    {
                    	my $job_list = make_job_list($execution);
                    	print $fh $job_list;
                        my $url = "https://" . $host . "/servlet/pluginDetail?executionId=" . $execution_id . "&recordId=" . $detail->{'id'};
				        my $request = HTTP::Request->new(GET => $url);
						my $response = $ua->request($request);
						
						if($response->is_success)
						{
							binmode $fh;
							print $fh $response->decoded_content;
						}
						else
						{
							die $response->status_line, "\n";
						}
				        
                    }
                }
            }
			close $fh;
        }
    }
    print "Wrote ".$output_file."\n";

	$netld->call("Security.logoutCurrentUser");

    return $last_end_time;
}

main();

File: Examples/Deprecated/Archive-Job-Execution-Outputs/Ruby/exportJobHistory.rb
#!/usr/bin/env ruby -W0

require 'jimson'
require 'uri'
require 'net/https'
require 'open-uri'
require 'zlib'
require 'optparse'
require 'inifile'
require 'date'

HOST='host'
USERNAME='username'
PASSWORD='password'
LAST_END_TIME='lastEndTime'

RAW_DEFLATE_ENCODING = -15

def main()
    options = {}
    OptionParser.new do |opts|
      opts.banner = "Usage: \n  exportJobHistory.rb [ <output file> ] [ <config file> ]"
    end.parse!
    
    outputFile = nil
    if ARGV.empty?
      outputFile = "JobHistory" + DateTime.parse(Time.now.to_s).strftime("%y%m%d-%H%M%S") + ".txt"
      configFile = './config.ini'
    else
      outputFile = ARGV[0]
      configFile = ARGV[1]
    end
    
    config = IniFile.load(configFile)

    config.each_section { |section|
        host = config[section][HOST]
        user = config[section][USERNAME]
        password = config[section][PASSWORD]
    
        if config[section][LAST_END_TIME]
            iniLastEndTime = config[section][LAST_END_TIME]
        else
            iniLastEndTime = 0
        end
        
        begin
            iniLastEndTime = exportJobHistory(outputFile, host, user, password, iniLastEndTime)
            
            if iniLastEndTime
              config[section][LAST_END_TIME] = iniLastEndTime
              config.write
            end 
        rescue => e
            puts e.to_s
        end
    }
end 

def millisToDate(millis)
    tmp = Time.at(millis / 1000).gmtime.strftime("%Y-%m-%d %H:%M:%S")
    return tmp 
end

def makeJobList(execution)
    startTime = execution['startTime'] / 1000
    endTime = execution['endTime'] / 1000
    
    jobList = "\n[ " + execution['status'].to_s + " / " +
            execution['jobName'].to_s + " / " +
            execution['managedNetwork'].to_s + " / " + 
            execution['executor'].to_s + " / S:" +
            Time.at(startTime).strftime('%Y-%m-%d %H:%M:%S') + " / E:" + 
            Time.at(endTime).strftime('%Y-%m-%d %H:%M:%S') + 
            " ]\n"
            
    return jobList
end

def exportJobHistory(outputFile, host, user, password, inilastEndTime)
  
    uri = URI.parse("https://" + host)
    https = Net::HTTP.new(uri.host, uri.port)
    https.use_ssl = true
    https.verify_mode = OpenSSL::SSL::VERIFY_NONE
    response = https.start { |https|
      https.get("/rest?j_username=" + user + "&j_password=" + password)
    }
    cookie = response['set-cookie'].split(';', 2)[0] 

    netld = Jimson::Client.new("https://" + host + "/rest?j_username=" + user + "&j_password=" + password, 
              "Cookie" => cookie)
 
    pageData = {'offset' => 0, 'pageSize' => 1000}
    pageData = netld['Scheduler.getExecutionData', pageData, 'endTime', 'True']

    file = nil
    lastExecuteTime = nil
    isFirst = 0

    pageData['executionData'].each { |execution|
      
        startTime = execution['startTime']
        lastEndTime = execution['endTime']
        endTime = millisToDate(lastEndTime)

        execution_id = execution['id']

        details = netld['Plugins.getExecutionDetails', execution_id]
          
        if details and execution['jobType'] == 'Script Tool Job' and lastEndTime > inilastEndTime

          if isFirst
            file = open(outputFile, 'w')
            lastExecuteTime = lastEndTime  
            isFirst = nil
          end
          
          details.each { |detail|
      
            jobList = makeJobList(execution)
            file.write(jobList)
            
            url = URI.parse("https://" + host + "/servlet/pluginDetail?executionId=" + execution_id.to_s + "&recordId=" + detail['id'].to_s)   
            response = https.start { |https|
              https.get(url, 
                {"Cookie" => cookie, "Accept-Encoding" => "deflate;q=0.6"}
              )
            }
            content = Zlib::Inflate.new(RAW_DEFLATE_ENCODING).inflate(response.body)
            
            file.write(content)
          }
        end
    }    

    if File.exist?(outputFile)
      file.close()
      puts "Wrote " + outputFile
    else
      puts "No Job History"
    end
    
    netld['Security.logoutCurrentUser']
    
    return lastExecuteTime
end

if __FILE__ == $0
  main()
end
